File: api/v1alpha1/conditions.go
// SPDX-FileCopyrightText: Jakob Naucke <jnaucke@redhat.com>
//
// SPDX-License-Identifier: MIT

package v1alpha1

const (
	InstalledCondition string = "Installed"
	InstalledReason string = "Installed"
	NotInstalledReasonNonUnique string = "NonUnique"
	NotInstalledReasonInstalling string = "Installing"
	NotInstalledReasonUninstalling string = "Uninstalling"

	KnownTrusteeAddressCondition string = "KnownTrusteeAddress"
	KnownTrusteeAddressReason string = "AddressFound"
	UnknownTrusteeAddressReason string = "NoAddressFound"

	CommittedCondition string = "Committed"
	CommittedReason string = "Committed"
	NotCommittedReasonComputing string = "Computing"
	NotCommittedReasonNoDigest string = "NoDigestGiven"
	NotCommittedReasonFailed string = "ComputationFailed"
)

File: operator/src/conditions.rs
// SPDX-FileCopyrightText: Jakob Naucke <jnaucke@redhat.com>
//
// SPDX-License-Identifier: MIT

use chrono::Utc;
use cocl_operator_lib::conditions::*;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::{Condition, Time};

pub fn condition_status(status: bool) -> String {
    match status {
        true => "True".to_string(),
        false => "False".to_string(),
    }
}

pub fn known_trustee_address_condition(known: bool, generation: Option<i64>) -> Condition {
    let err = "No publicTrusteeAddr specified. Components can deploy, \
               but register-server will not be able to point to Trustee until you add an address";
    let (reason, message) = match known {
        true => (KNOWN_TRUSTEE_ADDRESS_REASON, ""),
        false => (UNKNOWN_TRUSTEE_ADDRESS_REASON, err),
    };
    Condition {
        type_: KNOWN_TRUSTEE_ADDRESS_CONDITION.to_string(),
        status: condition_status(known),
        reason: reason.to_string(),
        message: message.to_string(),
        last_transition_time: Time(Utc::now()),
        observed_generation: generation,
    }
}

pub fn installed_condition(reason: &str, generation: Option<i64>) -> Condition {
    Condition {
        type_: INSTALLED_CONDITION.to_string(),
        status: condition_status(reason == INSTALLED_REASON),
        reason: reason.to_string(),
        message: match reason {
            NOT_INSTALLED_REASON_NON_UNIQUE => {
                "Another ConfidentialCluster definition was detected. \
                 Only one at a time is supported."
            }
            NOT_INSTALLED_REASON_INSTALLING => "Installation is in progress",
            NOT_INSTALLED_REASON_UNINSTALLING => "Uninstalling",
            _ => "",
        }
        .to_string(),
        last_transition_time: Time(Utc::now()),
        observed_generation: generation,
    }
}
